testmodule/providers/jsonrpc/testJsonRpcSubscriptions.nim

```diff
@@ -89,3 +94,58 @@ suite "HTTP polling subscriptions":
     await client.close()
 
   subscriptionTests(subscriptions, client)
+
+suite "HTTP polling subscriptions - filter not found":
+
+  var subscriptions: JsonRpcSubscriptions
+  var client: RpcHttpClient
+  var mockServer: MockRpcHttpServer
+
+  setup:
+    mockServer = MockRpcHttpServer.new()
+    mockServer.start()
+
+    client = newRpcHttpClient()
+    await client.connect("http://" & $mockServer.localAddress()[0])
+
+    subscriptions = JsonRpcSubscriptions.new(client,
+                                             pollingInterval = 100.millis)
+    subscriptions.start()
+
+  teardown:
+    await subscriptions.close()
+    await client.close()
+    await mockServer.stop()
+
+  test "filter not found error recreates filter":
+    let filter = EventFilter(address: Address.example, topics: @[array[32, byte].example])
+    let emptyHandler = proc(log: Log) = discard
+
+    check mockServer.newFilterCounter == 0
+    let jsonId = await subscriptions.subscribeLogs(filter, emptyHandler)
+    let id = string.fromJson(jsonId).tryGet
+    check mockServer.newFilterCounter == 1
+
+    await sleepAsync(300.millis)
+    mockServer.invalidateFilter(id)
+    await sleepAsync(300.millis)
+    check mockServer.newFilterCounter == 2
+
+  test "recreated filter can be still unsubscribed using the original id":
+    let filter = EventFilter(address: Address.example, topics: @[array[32, byte].example])
+    let emptyHandler = proc(log: Log) = discard
+
+    check mockServer.newFilterCounter == 0
+    let jsonId = await subscriptions.subscribeLogs(filter, emptyHandler)
+    let id = string.fromJson(jsonId).tryGet
+    check mockServer.newFilterCounter == 1
+
+    await sleepAsync(300.millis)
+    mockServer.invalidateFilter(id)
+    check eventually mockServer.newFilterCounter == 2
+    check mockServer.filters[id] == false
+    check mockServer.filters.len() == 2
+    await subscriptions.unsubscribe(jsonId)
+    check mockServer.filters.len() == 1
```

Comment:
Could maybe expose the subscription id mapping and check that they are indeed different?

Reply:
Well that mapping is not on the mock but in the `subscriptions` and there I would prefer to keep it internal. But I can assert that the only key in the `filters` is not the one that was originally returned which should give us the guarantee that "they are different".